### PR TITLE
[#ICOM-706] third party message validation

### PR DESCRIPTION
**Pull request by acuiuli**
#### List of Changes

- [new has_remote_content flag logic](https://github.com/pagopa/io-backend/pull/1079/commits/5507e91ca0ce22a13651c745bf159ce2f4a4c265)
- [fixed third party message validation logic and tests](https://github.com/pagopa/io-backend/pull/1079/commits/95541eb86cbc3df002ef064c4614875581caae84)
- [fixed code and test readability for the third party message validation](https://github.com/pagopa/io-backend/pull/1079/commits/c4118dd7ddf28a69447259b323db1648fc6d2772)

#### Motivation and Context

This pr adds the validation for the third party message received by the third party based on the flags has_attachments and has_remote_content. In particular it's important to check that if the third party set the flag to true than it must send the corresponding data in the getThirdPartyMessage api. There is an exception for PN since it is managed with different specs and different api.

#### How Has This Been Tested?

Unit test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/services/newMessagesService.ts [hunk 1]
```diff
@@ -541,11 +552,54 @@ export default class NewMessagesService {
...
           )
         )
       )
     );
 
+  private readonly validateThirdPartyMessageResponse = (
+    message: MessageWithThirdPartyData,
+    response: ThirdPartyMessageDetails
+  ): TE.TaskEither<IResponseErrorBadGateway, ThirdPartyMessage> => {
+    if (message.sender_service_id === PN_SERVICE_ID) {
```
Comment (gunzip) [inline src/services/newMessagesService.ts:567]: may we have a comment about this conditional statement here, expanding why PN is threated differently.

Comment (acuiuli) [reply src/services/newMessagesService.ts:568]: solved here [added comment for PN check in third party validation](https://github.com/pagopa/io-backend/pull/1079/commits/057af95cd30ea6d43003c3d774323a0d84326864)

**Review by Garma00 — APPROVED** [review]
lgtm

**Review by michaeldisaro — APPROVED** [review]

File: src/services/newMessagesService.ts [hunk 2]
```diff
@@ -541,11 +552,55 @@ export default class NewMessagesService {
...
         )
       )
     );
 
+  private readonly validateThirdPartyMessageResponse = (
+    message: MessageWithThirdPartyData,
+    response: ThirdPartyMessageDetails
+  ): TE.TaskEither<IResponseErrorBadGateway, ThirdPartyMessage> => {
+    // PN does not need this validation because it is managed completely in a different way with different specs
+    if (message.sender_service_id === PN_SERVICE_ID) {
```
Comment (gunzip) [inline src/services/newMessagesService.ts:568]: cc @lucacavallaro

File: src/services/newMessagesService.ts [hunk 3]
```diff
@@ -541,11 +552,80 @@ export default class NewMessagesService {
...
+      return TE.of(response);
+    }
+    if (
```
Comment (michaeldisaro) [inline src/services/newMessagesService.ts:571]: I'd separate each of these cases and put a comment on top of them to explain the check with common words.

File: src/services/newMessagesService.ts [hunk 4]
```diff
@@ -541,11 +552,80 @@ export default class NewMessagesService {
...
+      message.content.third_party_data.has_attachments &&
+      (!response.attachments || response.attachments.length === 0)
+    ) {
+      return TE.left(
+        ResponseErrorBadGateway(
+          InvalidThirdPartyMessageTypeEnum.ATTACHMENTS_NOT_PRESENT
+        )
+      );
+    }
+    if (
```
Comment (michaeldisaro) [inline src/services/newMessagesService.ts:581]: ditto

File: src/services/newMessagesService.ts [hunk 5]
```diff
@@ -541,11 +552,80 @@ export default class NewMessagesService {
...
+      message.content.third_party_data.has_remote_content &&
+      !response.details
+    ) {
+      return TE.left(
+        ResponseErrorBadGateway(
+          InvalidThirdPartyMessageTypeEnum.REMOTE_CONTENT_NOT_PRESENT
+        )
+      );
+    }
+    if (
```
Comment (michaeldisaro) [inline src/services/newMessagesService.ts:591]: ditto

File: src/services/newMessagesService.ts [hunk 6]
```diff
@@ -541,11 +552,80 @@ export default class NewMessagesService {
...
+      message.content.third_party_data.has_remote_content &&
+      !MessageBodyMarkdown.is(response.details.markdown)
+    ) {
+      return TE.left(
+        ResponseErrorBadGateway(
+          InvalidThirdPartyMessageTypeEnum.MARKDOWN_VALIDATION_ERROR
+        )
+      );
+    }
+    if (
```
Comment (michaeldisaro) [inline src/services/newMessagesService.ts:601]: ditto

File: src/services/__tests__/newMessagesService.test.ts [hunk 7]
```diff
@@ -823,20 +841,154 @@ describe("MessageService#getThirdPartyMessage", () => {
...
+  it("should return error Bad Gateway MARKDOWN_VALIDATION_ERROR if ThirParty service client returns an invalid markdown when the flag has_remote_content is true", async () => {
+    mockGetTPMessageFromExternalService.mockImplementationOnce(async (_messageId) =>
+      t.success({
+        status: 200,
+        headers: {},
+        value: {
+          ...aThirdPartyMessage,
+          details: {
+            ...aThirdPartyMessage.details,
+            markdown: "invalid"
```
Comment (michaeldisaro) [inline src/services/__tests__/newMessagesService.test.ts:912]: Maybe we should explain WHY it's invalid. I know that it should be over 80 chars, BUT other devs could have to search this "insight".

Comment (michaeldisaro) [reply src/services/__tests__/newMessagesService.test.ts:912]: Maybe we can refactor by extracting a `const invalidMarkdown = 'invalid'` and comment above that a valid markdown must be between 80 and 10000 chars. Or so.

File: src/services/__tests__/newMessagesService.test.ts [hunk 8]
```diff
@@ -823,20 +841,154 @@ describe("MessageService#getThirdPartyMessage", () => {
...
+  it("should return error Bad Gateway SUBJECT_VALIDATION_ERROR if ThirParty service client returns an invalid subject when the flag has_remote_content is true", async () => {
+    mockGetTPMessageFromExternalService.mockImplementationOnce(async (_messageId) =>
+      t.success({
+        status: 200,
+        headers: {},
+        value: {
+          ...aThirdPartyMessage,
+          details: {
+            ...aThirdPartyMessage.details,
+            subject: "invalid"
```
Comment (michaeldisaro) [inline src/services/__tests__/newMessagesService.test.ts:953]: ditto

**Review by michaeldisaro — CHANGES_REQUESTED** [review]

**Review by michaeldisaro — APPROVED** [review]

**Comment by michaeldisaro** [comment]
We have taken a task about refactoring unit tests.

**Pull request merged**